src/SessionSetup/include/SessionSetup/RetrieveInstancesWidget.h

```diff
@@ -23,15 +26,31 @@ class RetrieveInstancesWidget : public QWidget {
   Q_OBJECT
 
  public:
-  explicit RetrieveInstancesWidget(orbit_ggp::Client* ggp_client, QWidget* parent = nullptr);
+  explicit RetrieveInstancesWidget(MainThreadExecutor* main_thread_executor,
+                                   RetrieveInstances* retreive_instances,
+                                   QWidget* parent = nullptr);
   ~RetrieveInstancesWidget() override;
 
+  void Start();
+
  signals:
+  void LoadingStarted();
+  void LoadingSuccessful(QVector<orbit_ggp::Instance>);
+  void LoadingFailed();
+  void InitialLoadingFailed();
   void FilterTextChanged(const QString& text);
 
  private:
+  void SetupStateMachine();
+  orbit_ggp::Client::InstanceListScope GetInstanceListScope();
+  void InitialLoad(const std::optional<orbit_ggp::Project>& remembered_project);
+  std::optional<orbit_ggp::Project> GetQSettingsProject();
```

Comment:
`[[nodiscard]] std::optional<orbit_ggp::Project> GetQSettingsProject() const`

Reply:
I actually moved it into the cpp as a free function instead